### PR TITLE
Add type hints into optuna/integration/chainermn.py

### PR DESCRIPTION
**Pull request by norihitoishida**
## Motivation
Please refer to the Issue #711 .

## Description of the changes
I added type hints into optuna/integration/chainermn.py

### CODE REVIEW

File: optuna/integration/chainermn.py
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py:44]: `Callable` and `float` can be accessible in both type check and run time. `ChainerMNTrial` defines later of the code, and I think we use a string here.

```suggestion
        self, func: Callable[["ChainerMNTrial", "CommunicatorBase"], float], comm: "CommunicatorBase"
```

File: optuna/integration/chainermn.py
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py:50]: I think the use of `Trial` does not cause loop reference.
```suggestion
    def __call__(self, trial: Trial) -> float:
```

File: optuna/integration/chainermn.py
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py:77]: ditto.
```suggestion
    def __init__(self, study: Study, comm: "CommunicatorBase") -> None:
```

File: optuna/integration/chainermn.py
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py:100]: We can also access `Callable` and `float` here.
```suggestion
        func: Callable[["ChainerMNTrial", "CommunicatorBase"], float],
```

File: optuna/integration/chainermn.py
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py:163]: I think `Trial` will not cause loop reference.
```suggestion
    def __init__(self, trial: Optional[Trial], comm: "CommunicatorBase") -> None:
```

File: optuna/integration/chainermn.py
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py:216]: ditto.
```suggestion
    def suggest_categorical(self, name: str, choices: Sequence[CategoricalChoiceType]) -> Any:
        def func() -> CategoricalChoiceType:
```

File: optuna/integration/chainermn.py
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py:281]: ditto.
```suggestion
    def distributions(self) -> Dict[str, BaseDistribution]:
        def func() -> Dict[str, BaseDistribution]:
```

File: optuna/integration/chainermn.py
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py:315]: Can we use `Callable` type instead of the string here?
```suggestion
    def _call_with_mpi(self, func: Callable) -> Any:
```

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Thank you for your PR as always. This code still uses `if type_checking.TYPE_CHECKING` not to access `Trial`, `Study`, `BaseDistribution`, and `CategoricalChoiceType`, but I think we can remove it.  Please check my comments.

**Comment by norihitoishida** [comment]
> Thank you for your PR as always. This code still uses `if type_checking.TYPE_CHECKING` not to access `Trial`, `Study`, `BaseDistribution`, and `CategoricalChoiceType`, but I think we can remove it. Please check my comments.

Thank you for your review, I agree with you!

File: optuna/integration/chainermn.py
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py:19]: `NOQA` is no longer necessary because we use it in the code. And also, please reorder the import lines related to `optuna` in alphabetical order.
```suggestion
from optuna.distributions import BaseDistribution  # NOQA
from optuna.distributions import CategoricalChoiceType  # NOQA
from optuna.study import Study  # NOQA
from optuna.trial import Trial  # NOQA
```

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Thank you for your swift action. It basically looks good to me except for the order of the import lines.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the PR! LGTM!

**Review by hvy — APPROVED** [review]
Thanks as always, LGTM!

**Comment by hvy** [comment]
Let's also just wait for @toshihikoyanase san's approval.

**Review by toshihikoyanase — APPROVED** [review]
LGTM. Thank you for your update.

@hvy Thank you for letting me know.

**Pull request merged**
